Make array_pad's $length warning less confusing

nielsdos opened the pull request:
The error message was wrong; it *is* possible to use a larger length. The error message should be clear it's about how much padding is required.

Valid example with length greater than 1048576:
```php
<?php
var_dump(array_pad(array_fill(0, 1048576 * 2), 1048576 * 3 - 1, 0));
?>
```

If we were to remove the `- 1` in the above code snippet such that the required padding length becomes too large, it would be especially confusing for developers using the `array_pad` function.

~~This patch also removes a redundant check on pad_size_abs: it can never be negative since it is the absolute value of pad_size.~~

EDIT: I changed the code to check the *new* array length instead of having an arbtrary limit. This also simplifies the error message.

Note: I targeted the master branch because I'm not sure how much confusion is considered a bug.

mvorisek commented on ext/standard/tests/array/array_pad.phpt:
```diff
@@ -84,4 +84,4 @@ array(4) {
   float(2)
 }
-array_pad(): Argument #2 ($length) must be less than or equal to 1048576
+array_pad(): Argument #2 ($length) is too large because it results in a padding of 1999997 elements, which is larger than the allowed 1048576 elements
```
there should be no arbitrary limit as `1048576`, it was introduced 20 years ago in https://github.com/php/php-src/commit/7192e8af1531a1ed5568c415b75796f626a7abb2

nielsdos replied:
I thought the same thing; that removing the limit would be the ideal situation (whilst making sure we don't overflow the array length's limit, but that's a way higher limit that can't be hit in practice).
I'm willing to make a PR to do that, but it depends on the maintainers' opinion.

arnaud-lb replied:
We can use `HT_MAX_SIZE` as new limit

arnaud-lb replied:
We should probably limit in terms of total array size rather than padding size. This would also make the error clearer, and simplify the checks, if you do it in the same PR.

nielsdos replied:
Sounds good. I'll make a new PR with that approach to replace this one soon-ish.

Girgias commented on ext/standard/tests/array/array_pad.phpt:
```diff
@@ -84,4 +84,4 @@ array(4) {
   float(2)
 }
-array_pad(): Argument #2 ($length) must be less than or equal to 1048576
+array_pad(): Argument #2 ($length) is too large because it results in a padding of 1999997 elements, which is larger than the allowed 1048576 elements
```
I know this is already the case, but could you spin out the error condition out into a new test file?

Girgias reviewed (commented):
Not exactly sure of the wording, but this is definitely an improvement (also removes a duplicate computation).

nielsdos commented on ext/standard/array.c:
```diff
@@ -4227,8 +4227,9 @@ PHP_FUNCTION(array_pad)
-		zend_argument_value_error(2, "must be less than or equal to 1048576");
+	num_pads = pad_size_abs - input_size;
+	if (num_pads > Z_L(1048576)) {
+		zend_argument_value_error(2, "is too large because it results in a padding of %" PRId64 " elements, which is larger than the allowed 1048576 elements", num_pads);
```
Ah okay, I didn't realise this. I'll change it later today, thanks!

arnaud-lb commented:
> This patch also removes a redundant check on pad_size_abs: it can never be negative since it is the absolute value of pad_size.

Actually this can happen due to undefined behavior, when the result of abs() can not be represented in the return type.

With ZEND_ABS(), and assuming two's complement, this happens with pad_size=INTMAX_MIN only. If sizeof(intmax_t) > sizeof(zend_long), the conversion is implementation defined as well.

We could check that pad_size > LONG_MIN before calling ZEND_ABS().

We should add a test with length=PHP_INT_MIN.

nielsdos commented:
> > This patch also removes a redundant check on pad_size_abs: it can never be negative since it is the absolute value of pad_size.
> 
> Actually this can happen due to undefined behavior, when the result of abs() can not be represented in the return type.
> 

Right. My compiler actually optimizes away the check because of the UB. So on my system the old code didn't actually perform the `pad_size_abs<0` check at all.

> With ZEND_ABS(), and assuming two's complement, this happens with pad_size=INTMAX_MIN only. If sizeof(intmax_t) > sizeof(zend_long), the conversion is implementation defined as well.
> 
> We could check that pad_size > LONG_MIN before calling ZEND_ABS().
> 
> We should add a test with length=PHP_INT_MIN.

I can take a look at testing this soon-ish.

nielsdos commented:
I removed the arbitrary limit, and perform the check first on `pad_size` now.
As `HT_MAX_SIZE` is small enough, there are no overflow concerns anymore with `ZEND_ABS` or with negation, so the new code is quite simple.

cmb69 commented:
> I removed the arbitrary limit, and perform the check first on `pad_size` now.

That looks reasonable, but I wonder why that 1048576 limit had been chosen. It has been introduced by https://github.com/php/php-src/commit/4b343a05628ed0150c2bdd239d59a689e8697e0b, but the commit message is not helpful.

arnaud-lb commented:
Found the original commit: https://github.com/php/php-src/commit/7e8039ef0dd359ea7e7886f3f482a30f47fd8b6a, and I think a related internals post: https://marc.info/?l=php-internals&m=104919385614541&w=2, but no details about the crash or the fix, except that it's an arbitrarily chosen limit.

The implementation of array_pad back then has nothing in common with today's.

arnaud-lb commented:
From reading the code of array_pad, it seems safe as long as `HT_MAX_SIZE` has the right value. Unfortunately, trying to create an array with nearly `HT_MAX_SIZE` elements will cause some assertion failures in zend_hash.c due to `HT_SIZE_TO_MASK()` evaluating to `0`.

Also, `HT_MAX_SIZE` is a confusing name: zend_hash.c checks that the size is strictly lower than `HT_MAX_SIZE`.

arnaud-lb commented:
`HT_MAX_SIZE` now has the right value, and zend_hash allows to create arrays with a size <= `HT_MAX_SIZE`.

@nielsdos could you adjust `array_pad_too_large_padding.phpt`? Checking with just PHP_INT_MIN and PHP_INT_MAX is enough.

I will merge after that.

nielsdos commented:
> `HT_MAX_SIZE` now has the right value, and zend_hash allows to create arrays with a size <= `HT_MAX_SIZE`.
> 
> @nielsdos could you adjust `array_pad_too_large_padding.phpt`? Checking with just PHP_INT_MIN and PHP_INT_MAX is enough.
> 
> I will merge after that.

Thank you, I've rebased and updated the test.

arnaud-lb commented:
Thank you!

nielsdos commented:
Thank you for the merge!
Just a quick question: doesn't this also need an entry in UPGRADING, because the documentation needs to be updated? (https://www.php.net/manual/en/function.array-pad mentions the 1048576 limit)

arnaud-lb commented:
Indeed! I've just added an UPGRADING entry

nielsdos commented:
Thank you!
